File: 03-stream-processing/python/exercise1/data_generator.py
import json
import os
import random
import time
import uuid
from typing import Dict

import pendulum as pendulum
from pyspark.sql import SparkSession

print('Creating topics')
topic_name = 'visits'
os.system(
    f'docker exec p1_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic {topic_name} --delete')
os.system(
    f'docker exec p1_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic {topic_name} --create --partitions 2')


def random_visit() -> Dict[str, str]:
    return {'value': json.dumps({'eventId': str(uuid.uuid4()), 'os': random.choice(["macOS", "Windows", "Linux"]),
                                 'visitedPage': random.choice(
                                     ["category-1", "category-1/product-1", "index.html", "contact.html"])})}


spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.0').getOrCreate()

generation_day = pendulum.now('UTC').subtract(days=10)
while pendulum.now('UTC') > generation_day:
    print('Sending data')
    visits_to_generate = spark.createDataFrame(data=[
        random_visit(), random_visit(), random_visit(), random_visit(),
        random_visit(), random_visit(), random_visit(), random_visit(),
        random_visit(), random_visit(), random_visit(), random_visit(),
        random_visit(), random_visit(), random_visit(), random_visit(),
        random_visit(), random_visit(), random_visit(), random_visit()
    ], schema='value STRING')

    visits_to_generate.write.format("kafka") \
        .option("topic", topic_name).option("kafka.bootstrap.servers", "localhost:29092") \
        .save()

    generation_day = generation_day.add(minutes=1)
    time.sleep(5)

File: 03-stream-processing/scala/src/main/scala/com/becomedataengineer/exercise1/DataGenerator.scala
package com.becomedataengineer.exercise1

import org.apache.spark.sql.SparkSession

import java.time.{LocalDateTime, ZoneId}
import scala.util.Try
import scala.sys.process._

object DataGenerator {

  def main(args: Array[String]): Unit = {
    println("Preparing topics")
    val topicName = "visits"
    Try {
      val cmdDelete = s"docker exec p1_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic ${topicName} --delete"
      val outputDelete = cmdDelete.!!
    }
    val cmdCreate = s"docker exec p1_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic ${topicName} --create --partitions 2"
    val outputCreate = cmdCreate.!!

    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()
    import sparkSession.implicits._
    val utc = ZoneId.of("UTC")

    val firstDayToGenerate = LocalDateTime.now(utc).minusDays(10)
    var generationTime = firstDayToGenerate
    while (generationTime.isBefore(LocalDateTime.now(utc))) {
      println("Sending records")
      Seq(
        UserVisit().asJson, UserVisit().asJson, UserVisit().asJson, UserVisit().asJson,
        UserVisit().asJson, UserVisit().asJson, UserVisit().asJson, UserVisit().asJson,
        UserVisit().asJson, UserVisit().asJson, UserVisit().asJson, UserVisit().asJson,
        UserVisit().asJson, UserVisit().asJson, UserVisit().asJson, UserVisit().asJson,
        UserVisit().asJson, UserVisit().asJson, UserVisit().asJson, UserVisit().asJson
      ).toDF("value").write.format("kafka")
        .option("topic", topicName)
        .option("kafka.bootstrap.servers", "localhost:29092")
        .save()

      generationTime = generationTime.plusMinutes(1L)
      Thread.sleep(5000L)
    }
  }

}